### DISCUSSION TITLE
How to communicate trace-id between backend and frontend ? (langchain+langserve integration)

### DISCUSSION THREAD

**Discussion by Sinnaeve** [langfuse/langfuse#1952] [Support]
I use langchain+langserve as backend. I wanted to allow frontend to send user feedbacks as score to langfuse.
But it's unclear to me how the frontend can get the correct trace-id.

**Is there a proper way to have backend and frontend use the same trace-id when using langchain+langserve ?**
I feel like either the backend have to send the trace-id to the front in the endpoint output, or the front must be able to send the trace-id to be used to the backend. But I'm not sure how to do that  correclty.
Any help would be much appreciated !

______

Here are my research:
In: https://langfuse.com/docs/sdk/typescript/guide-web
it is stated:
While integrating Langfuse, it is important to either include the Langfuse Ids in the response to the frontend or to use an own id as the trace id which is available in both backend and frontend.

In discord someone managed it be re-writing the CallbackHandler it seems.
https://discord.com/channels/1111061815649124414/1211795284665831464/1223303914816929933

for my part:
I thought I might be able to use RunnableParallel from langchain to get current trace id as output:
````
final_chain= RunnableParallel(passed=chain, traceid=lambda x: langfuse_handler.get_trace_id())
````
But I'm not sure if my method would work with concurrent calls to the endpoint as I would have a single langfuse_handler object for mutliple calls.

**Comment by marcklingen** [comment]
Hi there, can you pass a separate langchain callback handler for every trace when using langserve?

**Comment by Sinnaeve** [comment]
I try using per request config modifier like this:

````
async def per_req_config_modifier(config: Dict, request: Request) -> Dict:
    """Modify the config for each request."""
    langfuse_session_id = config["configurable"]["langfuse_session_id"]
    langfuse_trace_id = config["configurable"]["langfuse_trace_id"]
    trace = langfuse_.trace(
            id=langfuse_trace_id,
            session_id=langfuse_session_id
        )
    if "callbacks" not in config:
        config["callbacks"] = []
    trace_handler = trace.get_langchain_handler()
    config["callbacks"].extend([trace_handler])
    return config

async def per_req_config_modifier(config: Dict, request: Request) -> Dict:
    """Modify the config for each request."""
    langfuse_session_id = config["configurable"]["langfuse_session_id"]
    langfuse_trace_id = config["configurable"]["langfuse_trace_id"]
    trace = langfuse_.trace(
            id=langfuse_trace_id,
            session_id=langfuse_session_id
        )
    if "callbacks" not in config:
        config["callbacks"] = []
    trace_handler = trace.get_langchain_handler()
    config["callbacks"].extend([trace_handler])
    return config

add_routes(app, agent, path="/agent",
           per_req_config_modifier=per_req_config_modifier)
````

But now I don't have trace name, input or output. I have the correct data in my 'generation' inside the trace. And I have output but not input in the Span contained in the trace that contains the generation.

Before I still didn't had input but I had output and trace name, using something like:
````
agent = agent.with_config( 
             config=RunnableConfig(
                 callbacks=[langfuse_handler],
             )
         )
````

____

I'm not to last langfuse version, so I will try to upgarde langchain/langserve/lanfuse to last version.

**Comment by Sinnaeve** [comment]
Ok, now I have updated to more recent versions:
langfuse: 2.28.2
langchain: 0.1.17
langserve: 0.1.1

I still don't have input and output for the trace (also the input and output are the good ones inside the Generation which is inside the trace.

I could manually add input when creating the trace with the following code, but that doesn't feel like it'sthe right way:

````
langfuse_= Langfuse(
    secret_key=os.environ['LANGFUSE_SECRET_KEY'],
    public_key=os.environ['LANGFUSE_PUBLIC_KEY'],
  host=os.environ['LANGFUSE_HOST'],
)

async def per_req_config_modifier(config: Dict, request: Request) -> Dict:

    """Modify the config for each request."""
    user = await get_current_active_user_from_request(request)
    langfuse_session_id = config["configurable"]["langfuse_session_id"]
    langfuse_trace_id = config["configurable"]["langfuse_trace_id"]
    trace = langfuse_.trace(
        id=langfuse_trace_id,
        session_id=langfuse_session_id,
        name='agent name',
        user_id=user.username,
        release=gitsha,
        input=request_data['input']['input'],
    )
    if "callbacks" not in config:
        config["callbacks"] = []
    trace_handler = trace.get_langchain_handler()
    config["callbacks"].extend([trace_handler])
    return config

add_routes(app, agent, path="/agent",
           per_req_config_modifier=per_req_config_modifier)
````

**Reply by marcklingen** [reply]
handlers created this way (`trace_handler = trace.get_langchain_handler()`) do not update the trace for which they are created as this is often undesirable when adding multiple runs to the same trace. We'll add an additional optional toggle for this this week

**Reply by quadcube** [reply]
> handlers created this way (`trace_handler = trace.get_langchain_handler()`) do not update the trace for which they are created as this is often undesirable when adding multiple runs to the same trace. We'll add an additional optional toggle for this this week

if we would like to update the trace's user_id in the per_req_config_modifier, how should we do it?

```
langfuse = Langfuse()
langfuse_handler = CallbackHandler()
config = RunnableConfig(callbacks=[langfuse_handler])

def per_req_config_modifier(config: Dict[str, Any], req: Request) -> Dict[str, Any]:
    config = config.copy()
    configurable = config.get("configurable", {})
   
    if "x-model-kwargs" in req.headers:
        configurable["model_kwargs"] = json.loads(req.headers["x-model-kwargs"])
    else:
        raise HTTPException(401, "No model arguments provided")
   
    user_id = configurable["model_kwargs"]["user"]
    config["configurable"] = configurable
    return config
 

chat_model = ChatOpenAI(
    model_kwargs = {}).configurable_fields(
        model_kwargs = ConfigurableField(id="model_kwargs")
    )

chain = some_prompt_template | chat_model | StrOutputParser()
add_routes(
    app,
    chain.with_types(input_type=InputChat).with_config(config),
    path="/some_chain",
    per_req_config_modifier=per_req_config_modifier,
)
```

Initially I thought we could get the trace or update the trace within the per_req_config_modifier function but it seems during execution of that function, the callback handler does not include the per_req_config_modifier since it's outside of scope.

**Reply by marcklingen** [reply]
Update on the above, you can now set the `update_parent` param on `get_langchain_handler()`
see reference: https://python.reference.langfuse.com/langfuse/client#StatefulTraceClient.get_langchain_handler

**Reply by marcklingen** [reply]
> if we would like to update the trace's user_id in the per_req_config_modifier, how should we do it?

I'd recommend to create a trace via the langfuse python sdk and then get a langchain callback handler for it as mentioned above. this is the most stable integration.

**Reply by quadcube** [reply]
> > if we would like to update the trace's user_id in the per_req_config_modifier, how should we do it?
> 
> I'd recommend to create a trace via the langfuse python sdk and then get a langchain callback handler for it as mentioned above. this is the most stable integration.

@marcklingen 
Thanks for the update!
I kinda solved it awhile back by implementing it as below. However I'm curious, what's the different between creating a trace and then getting the callback handler vs initializing a CallbackHandler() as below?

```
langfuse = Langfuse()

def per_req_config_modifier(config: Dict[str, Any], req: Request) -> Dict[str, Any]:
    config = config.copy()
    configurable = config.get("configurable", {})
   
    if "x-model-kwargs" in req.headers:
        configurable["model_kwargs"] = json.loads(req.headers["x-model-kwargs"])
    else:
        raise HTTPException(401, "No model arguments provided")

    if "callbacks" not in config:
        config["callbacks"] = []
   
    langfuse_handler = CallbackHandler(
        user_id=configurable["model_kwargs"]["user"],
        sessionI_id=configurable["session_id"] if "session_id" in configurable else None,
    )

    config["callbacks"].extend([langfuse_handler])

    config["configurable"] = configurable
    return config
 

chat_model = ChatOpenAI(
    model_kwargs = {}).configurable_fields(
        model_kwargs = ConfigurableField(id="model_kwargs")
    )

chain = some_prompt_template | chat_model | StrOutputParser()
add_routes(
    app,
    chain.with_types(input_type=InputChat).with_config(config),
    path="/some_chain",
    per_req_config_modifier=per_req_config_modifier,
)
```

**Reply by marcklingen** [reply]
Both approaches work, I’d recommend creating the trace independently when wanting to manipulate it further or to access the traceId concurrency-safe.

**Reply by chenxin6** [reply]
> > if we would like to update the trace's user_id in the per_req_config_modifier, how should we do it?
> 
> I'd recommend to create a trace via the langfuse python sdk and then get a langchain callback handler for it as mentioned above. this is the most stable integration.

@marcklingen I have tried this method like this.

```
from langfuse import Langfuse
langfuse_object = Langfuse()
trace = langfuse_object.trace(id=get_trace_id()) # The method get_trace_id can obtain the trace id of that API call.

async for event in chain.astream_events(
            {"messages": lc_messages},
            version="v1",
            config={"callbacks": [trace.get_langchain_handler(update_parent=True)]}  # type: ignore
):
    .............
    .............
```

But I get two traces like this. The trace id in red box is equal to the trace id of that API call. But the second trace `cc6f8b65` is not.
![image](https://github.com/user-attachments/assets/38b8b907-a133-49b3-9b9c-587212f993bd)

In the official trace demo, I find those two traces should be merged into one trace.
https://cloud.langfuse.com/project/cloramnkj0002jz088vzn1ja4/traces/4fcd93e3-79f2-474a-8e25-0e21c616249a
![image](https://github.com/user-attachments/assets/868757b6-89a0-4a7e-9f34-1c701f6b6dea)

**Reply by marcklingen** [reply]
@chenxin6, do you also import openai from langfuse?

**Reply by chenxin6** [reply]
> @chenxin6, do you also import openai from langfuse?

no. In this file, I only use langchain. Below are more code.

```
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI

def add_langfuse_trace(name: str, version: str):
    def add_langfuse_trace_wrapper(func):
        @observe()
        @wraps(func)
        def add_langfuse_trace_inner(*args, **kwargs):
            user: UserResponse = kwargs.get("user") # type: ignore
            param_dict = {
                "name": name,
                "version": version,
                "user_id": user.user_id,
                "input": {**kwargs},
                "metadata": {
                    "user_email": user.user_email,
                    "app_type": user.app_type,
                }
            }
            langfuse_context.update_current_trace(**param_dict)
            return func(*args, **kwargs)

        @wraps(func)
        def add_langfuse_trace_output(*args, **kwargs):
            kwargs["langfuse_observation_id"] = get_trace_id()
            return add_langfuse_trace_inner(*args, **kwargs)
        return add_langfuse_trace_output
    return add_langfuse_trace_wrapper

def create_message_service_chat(llm: ChatOpenAI):
    prompt = ChatPromptTemplate.from_messages(
        [
            MessagesPlaceholder(variable_name="messages"),
        ]
    ).with_config(tags=["prompt"])

    chain = prompt | llm

    return chain

@add_langfuse_trace(name="my_func", version="0")
def my_func(**kwargs):
    chain = create_message_service_chat(llm=ChatOpenAI(
        base_url=base_url,
        api_key=api_key,
        model=model_name,
        temperature=temperature,
        max_tokens=max_tokens
    ))

    from langfuse import Langfuse

    langfuse_object = Langfuse()
    trace = langfuse_object.trace(id=get_trace_id())

    async for event in chain.astream_events(
        {"messages": lc_messages},
        version="v1",
        config={"callbacks": [trace.get_langchain_handler(update_parent=True)]}  # type: ignore
    ):
        .............
        .............
```

**Reply by chenxin6** [reply]
@marcklingen I have come up with a solution based on above discussion.

```
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI

def add_langfuse_trace(name: str, version: str):
    def add_langfuse_trace_wrapper(func):
        @observe()
        @wraps(func)
        def add_langfuse_trace_inner(*args, **kwargs):
            user: UserResponse = kwargs.get("user") # type: ignore
            param_dict = {
                "name": name,
                "version": version,
                "user_id": user.user_id,
                "input": {**kwargs},
                "metadata": {
                    "user_email": user.user_email,
                    "app_type": user.app_type,
                }
            }
            langfuse_context.update_current_trace(**param_dict)
            return func(*args, **kwargs)

        @wraps(func)
        def add_langfuse_trace_output(*args, **kwargs):
            kwargs["langfuse_observation_id"] = get_trace_id()
            return add_langfuse_trace_inner(*args, **kwargs)
        return add_langfuse_trace_output
    return add_langfuse_trace_wrapper

def create_message_service_chat(llm: ChatOpenAI):
    prompt = ChatPromptTemplate.from_messages(
        [
            MessagesPlaceholder(variable_name="messages"),
        ]
    ).with_config(tags=["prompt"])

    chain = prompt | llm

    return chain

@add_langfuse_trace(name="my_func", version="0")
def my_func(**kwargs):
    chain = create_message_service_chat(llm=ChatOpenAI(
        base_url=base_url,
        api_key=api_key,
        model=model_name,
        temperature=temperature,
        max_tokens=max_tokens,
        model_kwargs={
            "trace_id": get_trace_id()
        }
    ))

    from langfuse.callback import CallbackHandler

    async for event in chain.astream_events(
        {"messages": lc_messages},
        version="v1",
        config={"callbacks": [CallbackHandler()], "run_id": get_trace_id()}  # type: ignore
    ):
        .............
        .............
```

**Reply by marcklingen** [reply]
thanks for sharing!
